### PR TITLE
Add authorization support based on roles and allowed actions

### PR DESCRIPTION
**Pull request by zhang-hua**
* Add Authorize annotations on controllers
* Add User Management client to get allowed actions from authentication service based on role info
* Validate allowed actions of current user against controller's required actions

### CODE REVIEW

File: iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/webservice/auth/IJwtValidation.java [hunk 1]
```diff
@@ -7,4 +7,5 @@
 @ImplementedBy(OpenIdConnectJwtValidation.class)
 public interface IJwtValidation {
     Boolean validateToken(String token) throws InvalidConfigurationException, ExternalDependencyException;
+    UserClaims getUserClaims(String token) throws NoAuthorizationException;
```
Comment (jillcary) [inline iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/webservice/auth/IJwtValidation.java]: Per feedback in https://github.com/Azure/pcs-config-dotnet/pull/95, change name to `NotAuthorizedException`

File: iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/webservice/auth/UserManagementClient.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @Override
+    public CompletionStage<List<String>> getAllowedActions(String userObjectId, List<String> roles) {
+        String url = String.format("%s/users/%s/allowedActions", this.userManagementServiceUrl, userObjectId);
+        return this.wsClient.url(url)
+            .post(Json.toJson(roles))
+            .handle((response, error) -> {
+                if (error != null) {
+                    String message = String.format("Fail to get allowed actions: %s", url);
```
Comment (jillcary) [inline iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/webservice/auth/UserManagementClient.java]: nit: "Failed"

File: iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/webservice/auth/ClientAuthConfig.java [hunk 3]
```diff
@@ -16,13 +17,15 @@
...
         Boolean authRequired,
+        String authServiceurl,
         String authType,
         HashSet<String> jwtAllowedAlgos,
         String jwtIssuer,
         String jwtAudience,
         Duration jwtClockSkew
     ) {
         this.authRequired = authRequired;
+        this.authServiceurl = authServiceurl;
```
Comment (jillcary) [inline iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/webservice/auth/ClientAuthConfig.java]: Nit: capitalize Url: `authServiceUrl`

File: iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/webservice/auth/IJwtValidation.java (hunk 1 above)
Comment (zhang-hua) [reply iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/webservice/auth/IJwtValidation.java]: make sense to me.

File: iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/webservice/auth/UserManagementClient.java (hunk 2 above)
Comment (zhang-hua) [reply iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/webservice/auth/UserManagementClient.java]: yep.

File: iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/webservice/auth/ClientAuthConfig.java (hunk 3 above)
Comment (zhang-hua) [reply iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/webservice/auth/ClientAuthConfig.java]: yep.

File: iothub-manager/README.md [hunk 4]
```diff
@@ -109,9 +113,9 @@ Steps using Eclipse Oxygen ("Eclipse for Java Developers" package):
   required by Eclipse to recognize the project.
 * Open Eclipse, and from the Welcome screen "Import" an existing project,
   navigating to the root folder of the project.
-* From the console run `sbt -jvm-debug 9999 "run 9005"` to start the project
+* From the console run `sbt -jvm-debug 9999 "run 9002"` to start the project
 * Test that the service is up and running pointing your browser to
-  http://127.0.0.1:9005/v1/status
+  http://127.0.0.1:900/v1/status
```
Comment (jillcary) [inline iothub-manager/README.md]: 9002?

Comment (zhang-hua) [reply iothub-manager/README.md]: oops.

File: iothub-manager/README.md [hunk 5]
```diff
@@ -60,10 +60,12 @@ for [Deploy the Azure services](https://docs.microsoft.com/azure/iot-suite/iot-s
...
 In order to run the service, some environment variables need to be created at least once. See specific instructions for IDE or command line setup below for more information. More information on environment variables [here](#configuration-and-environment-variables).
 
+* `PCS_AUTH_WEBSERVICE_URL` = http://localhost:9001/v1
+    * The url for the [Authentication microservice](https://github.com/Azure/pcs-auth-dotnet) from [Setup Dependencies](#setup-dependencies)
 * `PCS_IOTHUB_CONNSTRING` = {your Azure IoT Hub connection string from [Deploy Azure Services](#deploy-azure-services)}
     *  More information on where to find your IoT Hub connection string [here][iothub-connstring-blog].
-* `PCS_CONFIG_WEBSERVICE_URL` = http://localhost:9005/v1
-    * The url for the [Config microservice](https://github.com/Azure/pcs-config-java) from [Setup Dependencies](#setup-dependencies)
+* `PCS_STORAGEADAPTER_WEBSERVICE_URL` = http://localhost:9022/v1
+    * The url for the [Config microservice](https://github.com/Azure/pcs-storage-adapter-java) from [Setup Dependencies](#setup-dependencies)
```
Comment (jillcary) [inline iothub-manager/README.md]: `The url for the [Storage adapter microservice]`

File: iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/webservice/auth/IUserManagementClient.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (molly-moen) [inline iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/webservice/auth/IUserManagementClient.java]: in dotnet we put UserManagementClient in services/external, any reason for the change?

File: iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/webservice/auth/Authorize.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
+package com.microsoft.azure.iotsolutions.iothubmanager.webservice.auth;
```
Comment (jillcary) [inline iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/webservice/auth/Authorize.java]: `// Copyright (c) Microsoft. All rights reserved.`

**Pull request merged**